### PR TITLE
Dhanmoni gsoc21 toppview

### PR DESCRIPTION
**Pull request by timosachsenberg**
# Description

supersedes https://github.com/OpenMS/OpenMS/pull/5489 and adds a button to switch between horizontal and vertical table layout
image

# Checklist:
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.
- Or click on the action: e.g., for clang-format linting

# Note:
- Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

# Advanced commands (admins / reviewer only):
- /rebase will try to rebase the PR on the current develop branch.
- /reformat (experimental) applies the clang-format style changes as additional commit
- setting the label NoJenkins will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -31,4 +31,7 @@ cmake-build-debug
 cmake-build-release
 cmake-build-*
 *.code-workspace
-.vs
\ No newline at end of file
+.vs
+contrib_win64_build
+CMakeSettings.json
+OpenMS_Win64
```
Comment (cbielow) [inline .gitignore:37]: none of these are generally applicable IMHO. 
Remove?

File: src/openms_gui/include/OpenMS/VISUAL/SequenceVisualizer.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  public slots:
+    // this method sets protein and peptide data to m_json_data_obj.
+    void setProteinPeptideDataToJsonObj(
+        const QString& accession_num, 
+        const QString& pro_seq, 
+        const QJsonArray& peptides_data);
+
+  private:
+    Ui::SequenceVisualizer* ui;
+    QJsonObject m_json_data_obj;
```
Comment (cbielow) [inline src/openms_gui/include/OpenMS/VISUAL/SequenceVisualizer.h:70]: should be `m_json_data_obj_`

File: src/openms_gui/include/OpenMS/VISUAL/SpectraIDViewTab.h [hunk 3]
```diff
@@ -90,22 +98,40 @@ namespace OpenMS
...
     QCheckBox* hide_no_identification_ = nullptr;
     QCheckBox* create_rows_for_commmon_metavalue_ = nullptr;
     TableView* table_widget_ = nullptr;
+    TableView* protein_table_widget_ = nullptr;
     QTableWidget* fragment_window_ = nullptr;
-    bool is_ms1_shown_ = false;
-  
+    QSplitter* tables_splitter_ = nullptr;
+    bool is_first_time_loading = true;
+    std::unordered_map<String, std::vector<const PeptideIdentification*>> protein_to_peptide_id_map;
```
Comment (cbielow) [inline src/openms_gui/include/OpenMS/VISUAL/SpectraIDViewTab.h:116]: underscores missing

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 4]
```diff
@@ -178,8 +178,21 @@ namespace OpenMS
     connect(&tab_bar_, &EnhancedTabBar::dropOnTab, this, &TOPPViewBase::copyLayer);
     box_layout->addWidget(&tab_bar_);
 
+    //Trigger updates only when the active subWindow index changes and update it
     connect(&ws_, &EnhancedWorkspace::subWindowActivated, [this](QMdiSubWindow* window) {
-      if (window != nullptr) /* 0 upon terminate */ updateBarsAndMenus(); 
+      Size s = 0;
+      int oldactiveSubwindowID = activeSubwindowID_;
```
Comment (cbielow) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:184]: why not store the pointer to the window directly?
This would safe us from iterating through the window list and one can simply do:
```
if (window == nullptr) return; /* 0 upon terminate */ 
if (active_subwindow_ != window)
{
  active_subwindow_ == window;
  updateBarsAndMenus();
}
```

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 5]
```diff
@@ -1488,6 +1501,10 @@ namespace OpenMS
   {
     if (!ws_.currentSubWindow())
     {
+      if (activeSubwindowID_ < (Size) ws_.subWindowList().size())
```
Comment (cbielow) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:1504]: i'm not sure what this is doing...
if the subwindow is currently empty, why return the "old" one?
What happens if the window was closed?

Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:1504]: I think it had something to do with focus.
If the plot area is not focussed, you get a nullptr for currentSubwindow.
But I think some other code depends on getting information from the currently visible (but not active)
spectra plot. 
Not sure if I explained that well.

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp (hunk 4 above)
Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:184]: I think we need it later to get the currently _visible_ plot.
See my other comment. I'm not 100% sure anymore. I was and still am a bit confused what Qt means with "active" and if the MDI case if different. Sounds like it in the Qt docu.

File: src/openms_gui/source/VISUAL/ICONS/sequence_viz.html [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            
+            if(!protein_seq.length){ 
+                async function fetchSeqDataJSON() {
+                  const response = await fetch('https://www.ebi.ac.uk/proteins/api/proteins/' + acc_num, {
+                        method:'GET',
+                        headers: {
+                            'Content-Type': 'application/json'
+                        }
+                  })
+                  if (!response.ok) {
```
Comment (cbielow) [inline src/openms_gui/source/VISUAL/ICONS/sequence_viz.html:65]: what if ebi website is down?

Then it throws an error. Ok.
But what does that mean for the interface? Is it a messagebox or does it close the app?

File: src/openms_gui/source/VISUAL/PlotCanvas.cpp [hunk 7]
```diff
@@ -967,7 +967,7 @@ namespace OpenMS
 
   LayerDataBase& LayerStack::getCurrentLayer()
   {
-    if (current_layer_ >= layers_.size())
+    if (current_layer_ >= layers_.size() || current_layer_ < 0)
```
Comment (cbielow) [inline src/openms_gui/source/VISUAL/PlotCanvas.cpp:970]: impossible. Its a size_t ...

File: src/openms_gui/source/VISUAL/ICONS/sequence_viz.html (hunk 6 above)
Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/ICONS/sequence_viz.html:65]: I think it leaves the "browser" window half loaded without the sequence and throws an error in the "website log" (i.e. in the "Chrome tools" or equivalent view)

File: src/openms_gui/include/OpenMS/VISUAL/SequenceVisualizer.h [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+  public slots:
+    // this method sets protein and peptide data to m_json_data_obj.
+    void setProteinPeptideDataToJsonObj(
+        const QString& accession_num, 
+        const QString& pro_seq, 
+        const QJsonArray& peptides_data);
+
+  private:
+    Ui::SequenceVisualizer* ui;
```
Comment (cbielow) [inline src/openms_gui/include/OpenMS/VISUAL/SequenceVisualizer.h:69]: and `ui_`

File: src/openms_gui/source/VISUAL/SpectraTreeTab.cpp [hunk 9]
```diff
@@ -182,12 +182,18 @@ namespace OpenMS
...
+  }
+
   void SpectraTreeTab::itemSelectionChange_(QTreeWidgetItem* current, QTreeWidgetItem* previous)
   {
     /*	test for previous == 0 is important - without it,
         the wrong spectrum will be selected after finishing
         the execution of a TOPP tool on the whole data */
-    if (current == nullptr || previous == nullptr)
+    // WUT????
+    if (current == nullptr)
```
Comment (cbielow) [inline src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: has this been verified?
The test for `previous` was there for a reason?!

File: src/openms_gui/include/OpenMS/VISUAL/SpectraTreeTab.h [hunk 10]
```diff
@@ -92,13 +95,15 @@ namespace OpenMS
     QLineEdit* spectra_search_box_ = nullptr;
     QComboBox* spectra_combo_box_ = nullptr;
     TreeView* spectra_treewidget_ = nullptr;
+    LayerDataBase* layer_ = nullptr;
```
Comment (cbielow) [inline src/openms_gui/include/OpenMS/VISUAL/SpectraTreeTab.h:98]: why is this needed?
It seems to be assigned only once in the cpp and never read from.

File: src/openms_gui/source/VISUAL/SpectraTreeTab.cpp [hunk 11]
```diff
@@ -306,6 +313,7 @@ namespace OpenMS
       clear();
       return;
     }
+    layer_ = layer;
```
Comment (cbielow) [inline src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:316]: member can be removed IMHO

File: src/openms_gui/source/VISUAL/SpectraTreeTab.cpp (hunk 9 above)
Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: Not really, but this cannot be the solution to whatever problem this is trying to solve.
With this, you cannot ever set a selection if nothing was selected previously.

File: src/openms_gui/source/VISUAL/TVIdentificationViewController.cpp [hunk 12]
```diff
@@ -309,14 +312,17 @@ namespace OpenMS
...
-      return;
+    if (widget_1D == nullptr) 
+    {
+      std::cout << "Current widget is nullptr" << std::endl;
+      return; 
     }
 
     // lambda which returns the current layer
     // (this needs to be reevaluated, since adding a layer can invalidate the reference/pointer due to realloc)
+    // TODO if this needs to be done like that (e.g. for thread safety), why not in the other 10 occurrences of getting the layer in this class??
```
Comment (cbielow) [inline src/openms_gui/source/VISUAL/TVIdentificationViewController.cpp:323]: remove that comment.
Why its done is written right above it...

File: src/openms_gui/source/VISUAL/SpectraTreeTab.cpp [hunk 13]
```diff
@@ -428,6 +438,7 @@ namespace OpenMS
       {
         // now, select and scroll down to item
         selected_item->setSelected(true);
+        spectra_treewidget_->setCurrentItem(selected_item);
```
[the remark this answers is not in the document]
Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:441]: No, especially for keyboard navigation. See here: https://felgo.com/doc/qt/model-view-programming/#current-item-and-selected-items

[the remark this answers is not in the document]
Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:441]: E.g. In our coloring, selecting makes it "gray" but actually making it the _current_ item makes it blue.

File: src/openms_gui/source/VISUAL/SpectraTreeTab.cpp (hunk 9 above)
Comment (cbielow) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: The problem was, if i remember correctly, that TOPPView crashed.
Hence the extra check.
So now it crashes again, given the right circumstances, but we can set a selection?
Seems like this needs some investigation...

File: src/openms_gui/source/VISUAL/SpectraTreeTab.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (cbielow) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:441]: ahh interesting.
I keep forgetting :)
A comment would be nice

File: src/openms_gui/source/VISUAL/SpectraTreeTab.cpp (hunk 9 above)
Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: Yes true, I think whatever made it crash should probably be fixed instead of restricting other functions just to be safe.

@timosachsenberg can you quickly check? You have it built already.

File: src/openms_gui/source/VISUAL/TVIdentificationViewController.cpp (hunk 12 above)
Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/TVIdentificationViewController.cpp:323]: I mean more the other occurrences in the file. If thread-safety is a concern, I guess the rest of the file(s) have the same problem, but they were accessing the layer directly.

File: src/openms_gui/source/VISUAL/SpectraTreeTab.cpp (hunk 9 above)
Comment (timosachsenberg) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: sure what exactly?

File: src/openms_gui/source/VISUAL/TVIdentificationViewController.cpp (hunk 12 above)
Comment (cbielow) [reply src/openms_gui/source/VISUAL/TVIdentificationViewController.cpp:323]: ahh.. classical misunderstanding :)

Its not a threading issue, but the fact that down below, e.g. line
https://github.com/OpenMS/OpenMS/pull/5720/files#diff-b7d3caaae7f5a9154973da5f9a1ba05f9f7693d8ee6fd7644fce2880f95c519fR372

will add a new layer, thus potentially invalidating a "static" pointer to the current_layer.

Maybe my comment should be improved...

File: src/openms_gui/source/VISUAL/SpectraTreeTab.cpp (hunk 11 above)
Comment (timosachsenberg) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:316]: it is used in   void SpectraTreeTab::updateIndexFromCurrentLayer()
  {
    spectra_treewidget_->setTreePosition(layer_->getCurrentSpectrumIndex());
  }

File: src/openms_gui/source/VISUAL/TVIdentificationViewController.cpp (hunk 12 above)
Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/TVIdentificationViewController.cpp:323]: Aaah I even changed it in another function just to be sure 😃 cause I never know where the bug might come from

File: src/openms_gui/source/VISUAL/SpectraTreeTab.cpp (hunk 9 above)
Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: Where the comment says

Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: Sorry on phone

Comment (timosachsenberg) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: Didn't crash now

Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: Still? Or worse? 😆

Comment (timosachsenberg) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: hehe no still :) the usual strange stuff in TOPPView

Comment (jpfeuffer) [reply src/openms_gui/source/VISUAL/SpectraTreeTab.cpp:196]: Ok yes, I remember still having issues as well.

File: src/openms_gui/include/OpenMS/VISUAL/SequenceVisualizer.h [hunk 14]
```diff
@@ -35,29 +35,44 @@
 #ifdef QT_WEBENGINEWIDGETS_LIB
 #pragma once
 
-// OpenMS_GUI config
+// OpenMS_GUI configg
```
Comment (timosachsenberg) [inline src/openms_gui/include/OpenMS/VISUAL/SequenceVisualizer.h:38]: typo

File: src/openms_gui/source/VISUAL/SequenceVisualizer.cpp [hunk 15]
```diff
@@ -49,29 +49,36 @@ namespace OpenMS
...
       const QJsonArray& pep_data)
   {
-    m_json_data_obj_["accession_num"] = accession_num;
-    m_json_data_obj_["protein_sequence_data"] = pro_seq;
-    m_json_data_obj_["peptides_data"] = pep_data;
+    QJsonObject j;
+    j["accession_num"] = accession_num;
+    j["protein_sequence_data"] = pro_seq;
+    j["peptides_data"] = pep_data;
+    backend_.m_json_data_obj_ = j;
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/SequenceVisualizer.cpp:81]: can this be a move? or not necessary because "QJsonObject is an implicitly shared class, and shares the data with the document it has been created from as long as it is not being modified."

File: src/openms_gui/source/VISUAL/TableView.cpp [hunk 16]
```diff
@@ -328,6 +328,20 @@ namespace OpenMS
 
   void TableView::resizeEvent(QResizeEvent* /*event*/)
   {
+    this->resizeColumnsToContents();
+
+    int widgetWidth = this->viewport()->size().width();
+    int tableWidth = 0;
+
+    for(int i = 0; i<this->columnCount(); ++i)
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/TableView.cpp:336]: spaces

File: src/openms_gui/source/VISUAL/TableView.cpp [hunk 17]
```diff
@@ -328,6 +328,20 @@ namespace OpenMS
...
+      tableWidth += this->horizontalHeader()->sectionSize(i); //sections already resized to fit all data
+
+    double scale = (double)widgetWidth/tableWidth;
+    if (scale > 1.)
+    {
+      for(int i = 0; i<this->columnCount(); ++i)
```
Comment (timosachsenberg) [inline src/openms_gui/source/VISUAL/TableView.cpp:342]: spaces and brackets

**Review by jpfeuffer — APPROVED** [review]

**Comment by timosachsenberg** [comment]
Thanks!
